src/Aspire.Dashboard/Model/Otlp/SpanWaterfallViewModel.cs

```diff
@@ -47,62 +45,48 @@ public string GetTooltip(List<OtlpApplication> allApplications)
+                matchedDescendents = [child, ..matchedChildDescendents];
+                return true;
             }
-            else if (!string.IsNullOrEmpty(OtlpHelpers.GetValue(span.Attributes, "db.system")))
-            {
-                var dbSystem = OtlpHelpers.GetValue(span.Attributes, "db.system");
+        }
 
-                return $"DATA {dbSystem} {span.Name}";
-            }
-            else if (!string.IsNullOrEmpty(OtlpHelpers.GetValue(span.Attributes, "rpc.system")))
+        matchedDescendents = null;
+        return false;
+
+        bool Filter(SpanWaterfallViewModel viewModel)
+        {
+            if (string.IsNullOrWhiteSpace(filter))
             {
-                var rpcSystem = OtlpHelpers.GetValue(span.Attributes, "rpc.system");
-                var rpcService = OtlpHelpers.GetValue(span.Attributes, "rpc.service");
-                var rpcMethod = OtlpHelpers.GetValue(span.Attributes, "rpc.method");
-
-                if (string.Equals(rpcSystem, "grpc", StringComparison.OrdinalIgnoreCase))
-                {
-                    var grpcStatusCode = OtlpHelpers.GetValue(span.Attributes, "rpc.grpc.status_code");
-
-                    var summary = $"RPC {rpcService}/{rpcMethod}";
-                    if (!string.IsNullOrEmpty(grpcStatusCode) && Enum.TryParse<StatusCode>(grpcStatusCode, out var statusCode))
-                    {
-                        summary += $" {statusCode}";
-                    }
-                    return summary;
-                }
-
-                return $"RPC {rpcService}/{rpcMethod}";
+                return true;
             }
-            else if (!string.IsNullOrEmpty(OtlpHelpers.GetValue(span.Attributes, "messaging.system")))
-            {
-                var messagingSystem = OtlpHelpers.GetValue(span.Attributes, "messaging.system");
-                var messagingOperation = OtlpHelpers.GetValue(span.Attributes, "messaging.operation");
-                var destinationName = OtlpHelpers.GetValue(span.Attributes, "messaging.destination.name");
 
-                return $"MSG {messagingSystem} {messagingOperation} {destinationName}";
-            }
+            return viewModel.Span.SpanId.Contains(filter, StringComparison.CurrentCultureIgnoreCase)
+                   || getResourceName(viewModel.Span.Source).Contains(filter, StringComparison.CurrentCultureIgnoreCase)
+                   || viewModel.Span.GetDisplaySummary().Contains(filter, StringComparison.CurrentCultureIgnoreCase)
+                   || viewModel.UninstrumentedPeer?.Contains(filter, StringComparison.CurrentCultureIgnoreCase) is true;
         }
 
-        return span.Name;
+        static IEnumerable<SpanWaterfallViewModel> GetWithDescendents(SpanWaterfallViewModel s)
+        {
+            yield return s;
+
+            foreach (var descendent in s.Children.SelectMany(GetWithDescendents))
+            {
+                yield return descendent;
+            }
```

Comment:
Using recursion here could create a lot of enumerable instances in highly nested traces. Could you write this so it doesn't use recursion and creates a max of 1 enumerable instance.